src/hooks/useCopyToClipboard.tsx

```diff
@@ -0,0 +1,29 @@
+import { useState } from 'react'
+
+export interface useCopyToClipboardProps {
+  timeout?: number
+}
+
+export function useCopyToClipboard({ timeout = 2000 }: useCopyToClipboardProps) {
+  const [isCopied, setIsCopied] = useState<boolean>(false)
+
+  const copyToClipboard = (value: string) => {
+    if (typeof window === 'undefined' || !navigator.clipboard?.writeText) {
+      return
+    }
+
+    if (!value) {
+      return
+    }
+
+    void navigator.clipboard.writeText(value).then(() => {
+      setIsCopied(true)
+
+      setTimeout(() => {
+        setIsCopied(false)
+      }, timeout)
```

Comment:
Looks like there's a race condition here where its possible to invoke multiple promises before the first one resolves, resulting in `setIsCopied(true)` being fired out of order from `setIsCopied(false)`.

To prevent this you can either put this whole callback into a react-query, or dedupe the promise with

```
const [isCopying, setIsCopying] = useState(false)
```

You'd then do this at the start of this callback
```
if (isCopying) return
setIsCopying(true)
```

Then in the `setTimeout`
```
setIsCopying(true)
```

Reply:
Also noting this has a different purpose to `isCopied` as we need to set `isCopying` before we invoke the `writeText` promise

Reply:
Okay I can add that. 

Just to confirm this is what we are doing in the actual component itself to make sure if they are in the `isCopied` state they can't call `copyToClipboard` again until the setTimeout is called and we reset isCopied back to false.

```  const handleCopyClick = useCallback(() => {
    if (isCopied) return
    copyToClipboard(value)
  }, [copyToClipboard, isCopied, value])```

So we would be adding this logic to the hook instead of making the consumer do it?